cartographer/mapping/internal/2d/normal_estimation_2d.cc

```diff
@@ -0,0 +1,104 @@
+  }
+  Eigen::Vector3f mean_normal = Eigen::Vector3f::Zero();
+  const Eigen::Vector3f estimation_point_to_observation =
+      sensor_origin - estimation_point;
+  for (size_t sample_point_index = sample_window_lower_bound;
+       sample_point_index < sample_window_upper_bound; ++sample_point_index) {
+    if (sample_point_index == estimation_point_index) continue;
+    const Eigen::Vector3f sample_point = returns[sample_point_index];
+    const Eigen::Vector3f tangent = estimation_point - sample_point;
+    Eigen::Vector3f sample_normal = {-tangent[1], tangent[0], 0.f};
+    // Ensure sample_normal points towards 'sensor_origin'.
+    if (sample_normal.dot(estimation_point_to_observation) < 0) {
+      sample_normal = -sample_normal;
+    }
+    sample_normal.normalize();
+    mean_normal += sample_normal;
+  }
+  return Angle(mean_normal);
+}
+}  // namespace
+
+proto::NormalEstimationOptions2D CreateNormalEstimationOptions2D(
+    common::LuaParameterDictionary* parameter_dictionary) {
+  proto::NormalEstimationOptions2D options;
+  options.set_num_normal_samples(
+      parameter_dictionary->GetInt("num_normal_samples"));
+  options.set_sample_radius(parameter_dictionary->GetDouble("sample_radius"));
+  CHECK_GT(options.num_normal_samples(), 0);
+  CHECK_GT(options.sample_radius(), 0.0);
+  return options;
+}
+
+// Estimates the normal for each 'return' in 'range_data'.
+// Assumes the angles in the range data returns are sorted with respect to
+// the orientation of the vector from 'origin' to 'return'.
+std::vector<float> EstimateNormals(
+    const sensor::RangeData& range_data,
+    const proto::NormalEstimationOptions2D& normal_estimation_options) {
+  std::vector<float> normals;
+  const size_t max_num_samples = normal_estimation_options.num_normal_samples();
+  const float sample_radius = normal_estimation_options.sample_radius();
+  for (size_t current_point = 0; current_point < range_data.returns.size();
+       ++current_point) {
+    const Eigen::Vector3f& hit = range_data.returns[current_point];
+    size_t sample_window_lower_bound = current_point;
+    for (; sample_window_lower_bound > 0 &&
+           current_point - sample_window_lower_bound < max_num_samples / 2 &&
+           (hit - range_data.returns[sample_window_lower_bound - 1]).norm() <
+               sample_radius;
+         --sample_window_lower_bound) {
+    }
+    size_t sample_window_upper_bound = current_point;
+    for (; sample_window_upper_bound < range_data.returns.size() - 1 &&
+           sample_window_upper_bound - current_point <
+               max_num_samples / 2 + max_num_samples % 2 &&
+           (hit - range_data.returns[sample_window_upper_bound + 1]).norm() <
+               sample_radius;
+         ++sample_window_upper_bound) {
+    }
+    const float normal_estimate = EstimateNormal(
+        range_data.returns, current_point, sample_window_lower_bound,
+        sample_window_upper_bound + 1, range_data.origin);
```

Comment:
In this function, upper_bound is the last value in range, in the other function, it has the same name and is the first value after the range.
(The latter one is the C++ default for iterators.)
Either rename here (to first and last or the like) or change logic here, perhaps logic even becomes easier and removes some +/- 1.

Reply:
Agree, adhering too iterator defaults now.